tools/Dashboard/DevHome.Dashboard/ComSafeWidgetObjects/ComSafeWidget.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using DevHome.Common.Extensions;
+using DevHome.Dashboard.Services;
+using Microsoft.UI.Xaml;
+using Microsoft.Windows.Widgets;
+using Microsoft.Windows.Widgets.Hosts;
+using Windows.Foundation;
+
+namespace DevHome.Dashboard.ComSafeWidgetObjects;
+
+/// <summary>
+/// Since Widgets are OOP COM objects, we need to wrap them in a safe way to handle COM exceptions
+/// that arise when the underlying OOP object vanishes. All Widgets should be wrapped in a
+/// ComSafeWidget and calls to the widget should be done through the ComSafeWidget.
+/// This class will handle the COM exceptions and get a new OOP Widget if needed.
+/// All APIs on the IWidget and IWidget2 interfaces are reflected here.
+/// </summary>
+public class ComSafeWidget
+{
+    public Widget OopWidget { get; set; }
+
+    // Not currently used.
+    public DateTimeOffset DataLastUpdated => throw new NotImplementedException();
+
+    public string DefinitionId { get; private set; }
+
+    public string Id { get; private set; }
+
+    // Not currently used.
+    public DateTimeOffset TemplateLastUpdated => throw new NotImplementedException();
+
+    private const int RpcServerUnavailable = unchecked((int)0x800706BA);
+    private const int RpcCallFailed = unchecked((int)0x800706BE);
+
+    public ComSafeWidget(Widget widget)
+    {
+        OopWidget = widget;
+        DefinitionId = widget.DefinitionId;
+        Id = widget.Id;
+        widget.WidgetUpdated += OopWidgetUpdated;
+    }
+
+    public event TypedEventHandler<ComSafeWidget, WidgetUpdatedEventArgs> WidgetUpdated = (_, _) => { };
+
+    private void OopWidgetUpdated(Widget sender, WidgetUpdatedEventArgs args)
+    {
+        WidgetUpdated.Invoke(this, args);
+    }
+
+    public async Task<string> GetCardTemplateAsync()
+    {
+        return await Task.Run(async () =>
+        {
+            try
+            {
+                return await OopWidget.GetCardTemplateAsync();
+            }
+            catch (COMException e) when (e.HResult == RpcServerUnavailable || e.HResult == RpcCallFailed)
+            {
+                await GetNewOopWidgetAsync();
+                return await OopWidget.GetCardTemplateAsync();

```

Comment:
The `OopWidget` in the `catch` block can also throw an exception. I suggest doing a retry mechanism instead where we attempt a couples times before giving up and returning a non-RPC exception (e.g. `null` or error status which the UI can react to it).

Reply:
Current implementation decreases the change of an unhandled RPC exception, but does not fully protect from it 😉
Here's an example on a retry mechanism using exponential backoff (feel free to implement a different strategy that is more appropriate for widgets): [WinGetRecovery.cs](https://github.com/microsoft/devhome/blob/main/tools/SetupFlow/DevHome.SetupFlow/Services/WinGet/WinGetRecovery.cs)

Reply:
This is a good catch. I would expect if a retry fails we should handle the failure gracefully and log it unless we have reason to think it would succeed a short time later.

Reply:
Right now, in this class, I think it makes sense only do one retry where we re-get the object and call the given API on it. The WidgetHost and WidgetCatalog objects are the ones that do the heavy lifting of bringing the widgetservice back up so we can get the objects from that process. Over there, we try more times, so over here I don't think we need to try as many times. We can always adjust in the future.